### PR TITLE
feat: add support for arrays for otel traces

### PR DESCRIPTION
**Pull request by Ziinc**
This PR adds support for otel array attributes, as needed by storage team.

Screenshot 2025-05-12 at 6 58 36 PM

### CODE REVIEW

File: test/support/test_utils_grpc.ex [hunk 1]
```diff
@@ -75,8 +77,90 @@ defmodule Logflare.TestUtilsGrpc do
...
+      array_of_strings,
+      array_of_numbers,
+      array_of_booleans
+    ]
+  end
+
+  # this is not supported by traces/metrics, only logs
+  # https://github.com/open-telemetry/opentelemetry-specification/pull/2888
+  # https://github.com/open-telemetry/opentelemetry-specification/issues/376
+  defp random_key_value(:array_of_kv) do
```
Comment (jonatanklosko) [inline test/support/test_utils_grpc.ex:98]: This is not used anywhere. Perhaps expected, I am just noting to double-check :)

File: test/logflare_grpc/trace/server_test.exs [hunk 2]
```diff
@@ -5,6 +5,11 @@ defmodule LogflareGrpc.Trace.ServerTest do
   alias Opentelemetry.Proto.Collector.Trace.V1.ExportTraceServiceResponse
   alias Opentelemetry.Proto.Collector.Trace.V1.TraceService.Stub
   alias Logflare.SystemMetrics.AllLogsLogged
+  require OpenTelemetry.Tracer
+
+  require Record
+  @fields Record.extract(:span_ctx, from_lib: "opentelemetry_api/include/opentelemetry.hrl")
+  Record.defrecordp(:span_ctx, @fields)
```
Comment (jonatanklosko) [inline test/logflare_grpc/trace/server_test.exs:12]: This doesn't seem to be used anywhere?

File: test/logflare/logs/otel_trace_test.exs [hunk 3]
```diff
@@ -9,6 +9,25 @@ defmodule Logflare.Logs.OtelTraceTest do
...
+      assert Enum.any?(Map.values(a), &is_binary/1)
+      assert Enum.any?(Map.values(a), &is_boolean/1)
+    end
+
+    test "array handling", %{
+      resource_spans: resource_spans,
+      source: source
+    } do
+      batch = OtelTrace.handle_batch(resource_spans, source)
+    end
```
Comment (jonatanklosko) [inline test/logflare/logs/otel_trace_test.exs:29]: This doesn't assert anything and `handle_batch` is already called above with the same arguments. I would expect this test to be either expanded or removed :)

File: lib/logflare/logs/otel_trace.ex [hunk 4]
```diff
@@ -95,7 +96,20 @@ defmodule Logflare.Logs.OtelTrace do
 
   defp extract_key_value(%KeyValue{key: key, value: nil}), do: {key, nil}
 
+  defp extract_key_value(%KeyValue{
+         key: key,
+         value: %AnyValue{value: {:array_value, %ArrayValue{values: values}}}
+       }) do
+    {key, Enum.map(values, &extract_key_value/1)}
```
Comment (jonatanklosko) [inline lib/logflare/logs/otel_trace.ex:103]: The `Enum.map` here actually extracts just values. Perhaps it's worth separating a `extract_value` function, like this:

```elixir
defp extract_key_value(%KeyValue{key: key, value: value}) do
  {key, extract_value(value)}
end

defp extract_value(%AnyValue{value: {:array_value, %ArrayValue{values: values}}}) do
  Enum.map(values, &extract_value/1)
end

defp extract_value(%AnyValue{value: {_type, value}}), do: value
defp extract_value(nil), do: nil
defp extract_value(value), do: value
```

I added the last clause, because currently you have `defp extract_key_value(v), do: v`, though I am not sure if it's actually necessary, is it ever something other than `%AnyValue{}`?

File: test/support/test_utils_grpc.ex (hunk 1 above)
Comment (Ziinc) [reply test/support/test_utils_grpc.ex:98]: yea i intend to use it again when we support logs 😄

File: test/logflare_grpc/trace/server_test.exs (hunk 2 above)
Comment (Ziinc) [reply test/logflare_grpc/trace/server_test.exs:12]: ah remnants from testing. my bad, removed.

File: test/logflare/logs/otel_trace_test.exs (hunk 3 above)
Comment (Ziinc) [reply test/logflare/logs/otel_trace_test.exs:29]: removed as it array is already covered in above test.

File: lib/logflare/logs/otel_trace.ex (hunk 4 above)
Comment (Ziinc) [reply lib/logflare/logs/otel_trace.ex:103]: checked the otel spec and seems like AnyValue only gets used for arbitrary data values (this situation), but it is a bit vague if this is always used in the protobuf payload. 
I'll keep the last clause as fallback in case there is an unknown edge case, should be clearer if it is needed once we have better tooling around the protobuf struct generation for tests, in current state it is very much observational as opposed to using the otel sdk internals directly to generate the pb struct.
